test/test_infcholesky.jl

```diff
@@ -10,10 +11,16 @@ using InfiniteLinearAlgebra, LinearAlgebra, BandedMatrices, ArrayLayouts, Test
     @test cholesky(S) \ b ≈ qr(S) \ b ≈ S \ b
 
     @testset "singularly perturbed" begin
+        # using Symmetric(BandedMatrix(...))
         ε = 0.0001
         S = Symmetric(BandedMatrix(0 => 2 .+ ε*(1:∞), 1=> -Ones(∞)))
         b = [1; zeros(∞)]
         @test cholesky(S) \ b ≈  S \ b
+        # using SymTridiagonal(...)
+        ε = 0.0001
+        S = LazyBandedMatrices.SymTridiagonal(2 .+ ε*(1:∞), -Ones(∞))
+        @test MemoryLayout(S) isa SymmetricBandedLayouts
+        @test cholesky(S) \ b ≈  S \ b
```

Comment:
looks like there are 2 issues with DiagonalLayout. First, `cholesky(D)` doesn't end up in the adaptive cholesky (whatever method exists in LinearAlgebra is picked instead and thus we get stuck. Second, calling `adaptivecholesky(D)` directly fails because the method expects `parent(D)` to be a banded matrix but it's a vector for diagonal matrices. I could just special case diagonal matrices in `adaptivecholesky` to not rely on `parent(D)`.

Reply:
Okay nevermind, there is an unrelate  to the bug I was working on issue with diagonal matrices more fundamentally. So even

```julia
julia> S = Symmetric(BandedMatrix(0 => 2 .+ (1:∞)))
ℵ₀×ℵ₀ Symmetric{Float64, BandedMatrix{Float64, LazyArrays.ApplyArray{Float64, 2, typeof(*), Tuple{Matrix{Int64}, LazyArrays.ApplyArray{Float64, 2, typeof(vcat), Tuple{LazyArrays.ApplyArray{Float64, 2, typeof(hcat), Tuple{Zeros{Float64, 2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}, InfiniteArrays.ReshapedArray{Int64, 2, InfiniteArrays.InfUnitRange{Int64}, Tuple{Int64, Infinities.InfiniteCardinal{0}}, Tuple{}}}}}}}}, InfiniteArrays.OneToInf{Int64}}} with indices OneToInf()×OneToInf():
 3.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅     ⋅     ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   …  
  ⋅   4.0   ⋅    ⋅    ⋅    ⋅    ⋅     ⋅     ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅   5.0   ⋅    ⋅    ⋅    ⋅     ⋅     ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅   6.0   ⋅    ⋅    ⋅     ⋅     ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅   7.0   ⋅    ⋅     ⋅     ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅    ⋅   8.0   ⋅     ⋅     ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   …  
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   9.0    ⋅     ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅   10.0    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
  ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅     ⋅   11.0   ⋅    ⋅    ⋅    ⋅    ⋅    ⋅    ⋅      
 ⋮                        ⋮                          ⋮                        ⋮    ⋱  

julia> cholesky(S)
Cholesky{Float64, InfiniteLinearAlgebra.AdaptiveCholeskyFactors{Float64, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, BandedMatrix{Float64, LazyArrays.ApplyArray{Float64, 2, typeof(*), Tuple{Matrix{Int64}, LazyArrays.ApplyArray{Float64, 2, typeof(vcat), Tuple{LazyArrays.ApplyArray{Float64, 2, typeof(hcat), Tuple{Zeros{Float64, 2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}, InfiniteArrays.ReshapedArray{Int64, 2, InfiniteArrays.InfUnitRange{Int64}, Tuple{Int64, Infinities.InfiniteCardinal{0}}, Tuple{}}}}}}}}, InfiniteArrays.OneToInf{Int64}}}}
U factor:
ℵ₀×ℵ₀ UpperTriangular{Float64, InfiniteLinearAlgebra.AdaptiveCholeskyFactors{Float64, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, BandedMatrix{Float64, LazyArrays.ApplyArray{Float64, 2, typeof(*), Tuple{Matrix{Int64}, LazyArrays.ApplyArray{Float64, 2, typeof(vcat), Tuple{LazyArrays.ApplyArray{Float64, 2, typeof(hcat), Tuple{Zeros{Float64, 2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}, InfiniteArrays.ReshapedArray{Int64, 2, InfiniteArrays.InfUnitRange{Int64}, Tuple{Int64, Infinities.InfiniteCardinal{0}}, Tuple{}}}}}}}}, InfiniteArrays.OneToInf{Int64}}}} with indices OneToInf()×OneToInf():
 ERROR: BoundsError: attempt to access 0-element UnitRange{Int64} at index [1]
Stacktrace:
  [1] throw_boundserror(A::UnitRange{Int64}, I::Int64)
    @ Base ./abstractarray.jl:744
  [2] getindex
    @ ./range.jl:915 [inlined]
  [3] _shift
    @ ~/.julia/packages/BandedMatrices/2mDGG/src/banded/BandedMatrix.jl:843 [inlined]
  [4] similar
    @ ~/.julia/packages/BandedMatrices/2mDGG/src/banded/BandedMatrix.jl:850 [inlined]
  [5] _convert_common_container
    @ ~/.julia/packages/BandedMatrices/2mDGG/src/banded/BandedMatrix.jl:128 [inlined]
  [6] convert(#unused#::Type{BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}}, M::SubArray{Float64, 2, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false})
    @ BandedMatrices ~/.julia/packages/BandedMatrices/2mDGG/src/banded/BandedMatrix.jl:137
  [7] convert
    @ ~/.julia/packages/BandedMatrices/2mDGG/src/banded/BandedMatrix.jl:147 [inlined]
  [8] materialize!(M::MulAdd{BandedMatrices.BandedRows{DenseColumnMajor}, BandedMatrices.BandedColumns{DenseColumnMajor}, BandedMatrices.BandedColumns{DenseColumnMajor}, Float64, Adjoint{Float64, SubArray{Float64, 2, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}}, SubArray{Float64, 2, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}, SubArray{Float64, 2, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}})
    @ BandedMatrices ~/.julia/packages/BandedMatrices/2mDGG/src/generic/matmul.jl:184
  [9] muladd!
    @ ~/.julia/packages/ArrayLayouts/kR7AR/src/muladd.jl:70 [inlined]
 [10] partialcholesky!(F::InfiniteLinearAlgebra.AdaptiveCholeskyFactors{Float64, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, BandedMatrix{Float64, LazyArrays.ApplyArray{Float64, 2, typeof(*), Tuple{Matrix{Int64}, LazyArrays.ApplyArray{Float64, 2, typeof(vcat), Tuple{LazyArrays.ApplyArray{Float64, 2, typeof(hcat), Tuple{Zeros{Float64, 2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}, InfiniteArrays.ReshapedArray{Int64, 2, InfiniteArrays.InfUnitRange{Int64}, Tuple{Int64, Infinities.InfiniteCardinal{0}}, Tuple{}}}}}}}}, InfiniteArrays.OneToInf{Int64}}}, n::Int64)
    @ InfiniteLinearAlgebra ~/Documents/Projects/InfiniteLinearAlgebra.jl/src/infcholesky.jl:35
 [11] colsupport
    @ ~/Documents/Projects/InfiniteLinearAlgebra.jl/src/infcholesky.jl:54 [inlined]
 [12] colsupport
    @ ~/.julia/packages/ArrayLayouts/kR7AR/src/triangular.jl:2 [inlined]
 [13] colsupport
    @ ~/.julia/packages/ArrayLayouts/kR7AR/src/memorylayout.jl:657 [inlined]
 [14] layout_replace_in_print_matrix
    @ ~/.julia/packages/ArrayLayouts/kR7AR/src/ArrayLayouts.jl:339 [inlined]
 [15] replace_in_print_matrix(A::UpperTriangular{Float64, InfiniteLinearAlgebra.AdaptiveCholeskyFactors{Float64, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, BandedMatrix{Float64, LazyArrays.ApplyArray{Float64, 2, typeof(*), Tuple{Matrix{Int64}, LazyArrays.ApplyArray{Float64, 2, typeof(vcat), Tuple{LazyArrays.ApplyArray{Float64, 2, typeof(hcat), Tuple{Zeros{Float64, 2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}, InfiniteArrays.ReshapedArray{Int64, 2, InfiniteArrays.InfUnitRange{Int64}, Tuple{Int64, Infinities.InfiniteCardinal{0}}, Tuple{}}}}}}}}, InfiniteArrays.OneToInf{Int64}}}}, i::Int64, j::Int64, s::String)
    @ ArrayLayouts ~/.julia/packages/ArrayLayouts/kR7AR/src/ArrayLayouts.jl:350
 [16] print_matrix_row(io::IOContext{Base.TTY}, X::AbstractVecOrMat, A::Vector{Tuple{Int64, Int64}}, i::Int64, cols::Vector{Int64}, sep::String, idxlast::Infinities.InfiniteCardinal{0})
    @ Base ./arrayshow.jl:120
 [17] axes_print_matrix_row (repeats 2 times)
    @ ~/.julia/packages/FillArrays/wIoUg/src/FillArrays.jl:679 [inlined]
 [18] print_matrix_row(io::IOContext{Base.TTY}, X::UpperTriangular{Float64, InfiniteLinearAlgebra.AdaptiveCholeskyFactors{Float64, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, BandedMatrix{Float64, LazyArrays.ApplyArray{Float64, 2, typeof(*), Tuple{Matrix{Int64}, LazyArrays.ApplyArray{Float64, 2, typeof(vcat), Tuple{LazyArrays.ApplyArray{Float64, 2, typeof(hcat), Tuple{Zeros{Float64, 2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}, InfiniteArrays.ReshapedArray{Int64, 2, InfiniteArrays.InfUnitRange{Int64}, Tuple{Int64, Infinities.InfiniteCardinal{0}}, Tuple{}}}}}}}}, InfiniteArrays.OneToInf{Int64}}}}, A::Vector{Tuple{Int64, Int64}}, i::Int64, cols::Vector{Int64}, sep::String, idxlast::Infinities.InfiniteCardinal{0})
    @ ArrayLayouts ~/.julia/packages/ArrayLayouts/kR7AR/src/ArrayLayouts.jl:362
 [19] _print_matrix(io::IOContext{Base.TTY}, X::AbstractVecOrMat, pre::String, sep::String, post::String, hdots::String, vdots::String, ddots::String, hmod::Int64, vmod::Int64, rowsA::InfiniteArrays.InfUnitRange{Int64}, colsA::InfiniteArrays.InfUnitRange{Int64})
    @ Base ./arrayshow.jl:252
 [20] print_matrix(io::IOContext{Base.TTY}, X::UpperTriangular{Float64, InfiniteLinearAlgebra.AdaptiveCholeskyFactors{Float64, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, BandedMatrix{Float64, LazyArrays.ApplyArray{Float64, 2, typeof(*), Tuple{Matrix{Int64}, LazyArrays.ApplyArray{Float64, 2, typeof(vcat), Tuple{LazyArrays.ApplyArray{Float64, 2, typeof(hcat), Tuple{Zeros{Float64, 2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}, InfiniteArrays.ReshapedArray{Int64, 2, InfiniteArrays.InfUnitRange{Int64}, Tuple{Int64, Infinities.InfiniteCardinal{0}}, Tuple{}}}}}}}}, InfiniteArrays.OneToInf{Int64}}}}, pre::String, sep::String, post::String, hdots::String, vdots::String, ddots::String, hmod::Int64, vmod::Int64)
    @ Base ./arrayshow.jl:171
 [21] print_matrix
    @ ./arrayshow.jl:171 [inlined]
 [22] print_array
    @ ./arrayshow.jl:358 [inlined]
 [23] show(io::IOContext{Base.TTY}, #unused#::MIME{Symbol("text/plain")}, X::UpperTriangular{Float64, InfiniteLinearAlgebra.AdaptiveCholeskyFactors{Float64, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, BandedMatrix{Float64, LazyArrays.ApplyArray{Float64, 2, typeof(*), Tuple{Matrix{Int64}, LazyArrays.ApplyArray{Float64, 2, typeof(vcat), Tuple{LazyArrays.ApplyArray{Float64, 2, typeof(hcat), Tuple{Zeros{Float64, 2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}, InfiniteArrays.ReshapedArray{Int64, 2, InfiniteArrays.InfUnitRange{Int64}, Tuple{Int64, Infinities.InfiniteCardinal{0}}, Tuple{}}}}}}}}, InfiniteArrays.OneToInf{Int64}}}})
    @ Base ./arrayshow.jl:399
 [24] show(io::IOContext{Base.TTY}, mime::MIME{Symbol("text/plain")}, C::Cholesky{Float64, InfiniteLinearAlgebra.AdaptiveCholeskyFactors{Float64, BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, BandedMatrix{Float64, LazyArrays.ApplyArray{Float64, 2, typeof(*), Tuple{Matrix{Int64}, LazyArrays.ApplyArray{Float64, 2, typeof(vcat), Tuple{LazyArrays.ApplyArray{Float64, 2, typeof(hcat), Tuple{Zeros{Float64, 2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}, InfiniteArrays.ReshapedArray{Int64, 2, InfiniteArrays.InfUnitRange{Int64}, Tuple{Int64, Infinities.InfiniteCardinal{0}}, Tuple{}}}}}}}}, InfiniteArrays.OneToInf{Int64}}}})
    @ LinearAlgebra ~/Documents/Backends/julia-1.9.0/share/julia/stdlib/v1.9/LinearAlgebra/src/cholesky.jl:562
 [25] (::REPL.var"#55#56"{REPL.REPLDisplay{REPL.LineEditREPL}, MIME{Symbol("text/plain")}, Base.RefValue{Any}})(io::Any)
    @ REPL ~/Documents/Backends/julia-1.9.0/share/julia/stdlib/v1.9/REPL/src/REPL.jl:276
 [26] with_repl_linfo(f::Any, repl::REPL.LineEditREPL)
    @ REPL ~/Documents/Backends/julia-1.9.0/share/julia/stdlib/v1.9/REPL/src/REPL.jl:557
 [27] display(d::REPL.REPLDisplay, mime::MIME{Symbol("text/plain")}, x::Any)
    @ REPL ~/Documents/Backends/julia-1.9.0/share/julia/stdlib/v1.9/REPL/src/REPL.jl:262
 [28] display(d::REPL.REPLDisplay, x::Any)
    @ REPL ~/Documents/Backends/julia-1.9.0/share/julia/stdlib/v1.9/REPL/src/REPL.jl:281
 [29] display(x::Any)
    @ Base.Multimedia ./multimedia.jl:340
 [30] #invokelatest#2
    @ ./essentials.jl:816 [inlined]
 [31] invokelatest
    @ ./essentials.jl:813 [inlined]
 [32] (::VSCodeServer.var"#66#70"{Bool, Bool, Bool, Module, String, Int64, Int64, String, VSCodeServer.ReplRunCodeRequestParams})()
    @ VSCodeServer ~/.vscode/extensions/julialang.language-julia-1.47.2/scripts/packages/VSCodeServer/src/eval.jl:199
 [33] withpath(f::VSCodeServer.var"#66#70"{Bool, Bool, Bool, Module, String, Int64, Int64, String, VSCodeServer.ReplRunCodeRequestParams}, path::String)
    @ VSCodeServer ~/.vscode/extensions/julialang.language-julia-1.47.2/scripts/packages/VSCodeServer/src/repl.jl:249
 [34] (::VSCodeServer.var"#65#69"{Bool, Bool, Bool, Module, String, Int64, Int64, String, VSCodeServer.ReplRunCodeRequestParams})()
    @ VSCodeServer ~/.vscode/extensions/julialang.language-julia-1.47.2/scripts/packages/VSCodeServer/src/eval.jl:155
 [35] hideprompt(f::VSCodeServer.var"#65#69"{Bool, Bool, Bool, Module, String, Int64, Int64, String, VSCodeServer.ReplRunCodeRequestParams})
    @ VSCodeServer ~/.vscode/extensions/julialang.language-julia-1.47.2/scripts/packages/VSCodeServer/src/repl.jl:38
 [36] (::VSCodeServer.var"#64#68"{Bool, Bool, Bool, Module, String, Int64, Int64, String, VSCodeServer.ReplRunCodeRequestParams})()
    @ VSCodeServer ~/.vscode/extensions/julialang.language-julia-1.47.2/scripts/packages/VSCodeServer/src/eval.jl:126
 [37] with_logstate(f::Function, logstate::Any)
    @ Base.CoreLogging ./logging.jl:514
 [38] with_logger
    @ ./logging.jl:626 [inlined]
 [39] (::VSCodeServer.var"#63#67"{VSCodeServer.ReplRunCodeRequestParams})()
    @ VSCodeServer ~/.vscode/extensions/julialang.language-julia-1.47.2/scripts/packages/VSCodeServer/src/eval.jl:225
 [40] #invokelatest#2
    @ ./essentials.jl:816 [inlined]
 [41] invokelatest(::Any)
    @ Base ./essentials.jl:813
 [42] macro expansion
    @ ~/.vscode/extensions/julialang.language-julia-1.47.2/scripts/packages/VSCodeServer/src/eval.jl:34 [inlined]
 [43] (::VSCodeServer.var"#61#62")()
    @ VSCodeServer ./task.jl:514
```

fails and errors, i.e. without any use of DiagonalLayout. Seems like a bug in the cholesky method itself for 0 super and sub diagonal data. For diagonal matrices isn't the cholesky decomp trivially `sqrt.(D)` ? should we special case this in adaptive cholesky?